File: NSIDC/MOD29.A2013196.1250.061.2021233075404.hdf.py
"""
This example code illustrates how to access and visualize a NSIDC MOD29
Level 2 HDF-EOS2 Swath file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python MOD29.A2013196.1250.061.2021233075404.hdf.py

The HDF files must be in your current working directory.

Tested under: Python 3.9.13 :: Miniconda
Last updated: 2024-08-07
"""

import os

import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.basemap import Basemap
from pyhdf.SD import SD, SDC

FILE_NAME = "MOD29.A2013196.1250.061.2021233075404.hdf"
hdf = SD(FILE_NAME, SDC.READ)

# Read dataset.
DATAFIELD_NAME = "Ice_Surface_Temperature"
data2D = hdf.select(DATAFIELD_NAME)
data = data2D[:, :].astype(np.float64)

# Read attributes.
attrs = data2D.attributes(full=1)
aoa = attrs["add_offset"]
add_offset = aoa[0]
fva = attrs["_FillValue"]
_FillValue = fva[0]
sfa = attrs["scale_factor"]
scale_factor = sfa[0]
ua = attrs["units"]
units = ua[0]
va = attrs["valid_range"]
valid_range = va[0]
lna = attrs["long_name"]
long_name = lna[0]

# Read lat and lon data from the matching geo-location file.
GEO_FILE_NAME = "MOD03.A2013196.1250.061.2017299150213.hdf"
hdf_geo = SD(GEO_FILE_NAME, SDC.READ)

# Read geolocation dataset from MOD03 product.
lat = hdf_geo.select("Latitude")
latitude = lat[:, :]
lon = hdf_geo.select("Longitude")
longitude = lon[:, :]

# Apply the attributes.
invalid = np.logical_or(data < valid_range[0], data > valid_range[1])
invalid = np.logical_or(invalid, data == _FillValue)
data[invalid] = np.nan
data = data * scale_factor + add_offset
datam = np.ma.masked_array(data, mask=np.isnan(data))

# Draw a southern polar stereographic projection using the low resolution
# coastline database.
m = Basemap(projection="spstere", resolution="l", boundinglat=-64, lon_0=180)
m.drawcoastlines(linewidth=0.5)
m.drawparallels(np.arange(-80.0, -59, 10.0))
m.drawmeridians(
    np.arange(-180.0, 179.0, 30.0), labels=[True, False, False, True]
)
m.pcolormesh(longitude, latitude, datam, latlon=True)
cb = m.colorbar()
cb.set_label(units)

basename = os.path.basename(FILE_NAME)
plt.title("{0}\n{1}".format(basename, long_name))
fig = plt.gcf()
pngfile = "{0}.py.png".format(basename)
fig.savefig(pngfile)

File: NSIDC/MOD29_A2013196_1250_061_2021233075404_hdf.m
%
%  This example code illustrates how to access and visualize NSIDC
%  MOD29 Level 2 HDF-EOS2 Swath file in MATLAB.
%
%  If you have any questions, suggestions, comments on this example, 
% please use the HDF-EOS Forum (http://hdfeos.org/forums). 
% 
%  If you would like to see an  example of any other NASA HDF/HDF-EOS
% data product that is not listed in the HDF-EOS  Comprehensive Examples 
% page (http://hdfeos.org/zoo), feel free to contact us at 
% eoshelp@hdfgroup.org or post it at the HDF-EOS Forum 
% (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%
%  $matlab -nosplash -nodesktop -r MOD29_A2013196_1250_061_2021233075404_hdf
%
% Tested under: MATLAB R2023b
% Last updated: 2024-08-07

import matlab.io.hdfeos.*
import matlab.io.hdf4.*


% Set file name.
FILE_NAME = 'MOD29.A2013196.1250.061.2021233075404.hdf';

% Open the HDF-EOS2 swath file.
file_id = sw.open(FILE_NAME, 'rdonly');

% Set swath name.   
SWATH_NAME = 'MOD_Swath_Sea_Ice';
   
% Open swath.
swath_id = sw.attach(file_id, SWATH_NAME);

% Read data from a data field.
DATAFIELD_NAME = 'Ice_Surface_Temperature';

% Read the dataset.
data = sw.readField(swath_id, DATAFIELD_NAME, [], [], []);

% Detach from the swath object.
sw.detach(swath_id);
sw.close(file_id);

% Read lat and lon data from the matching geo-location file.
GEO_FILE_NAME='MOD03.A2013196.1250.061.2017299150213.hdf';
SWATH_NAME='MODIS_Swath_Type_GEO';

% Open the HDF-EOS2 swath file.
file_id = sw.open(GEO_FILE_NAME, 'rdonly');

% Open swath.
swath_id = sw.attach(file_id, SWATH_NAME);

% Read lat and lon.
lon = sw.readField(swath_id, 'Longitude', [], [], []);
lat = sw.readField(swath_id, 'Latitude', [], [], []);

% Detach swath.
sw.detach(swath_id);
sw.close(file_id);

% Convert the data to double type for plot.
data=double(data);
lon=double(lon);
lat=double(lat);

% Read attributes from the data field.
SD_id = sd.start(FILE_NAME, 'rdonly');
DATAFIELD_NAME = 'Ice_Surface_Temperature';
sds_index = sd.nameToIndex(SD_id, DATAFIELD_NAME);
sds_id = sd.select(SD_id, sds_index);

% Read fill value attribute.
fillvalue_index = sd.findAttr(sds_id, '_FillValue');
fillvalue = sd.readAttr(sds_id, fillvalue_index);

% Read long_name attribute.
long_name_index = sd.findAttr(sds_id, 'long_name');
long_name = sd.readAttr(sds_id, long_name_index);

% Read valid range.
valid_range_index = sd.findAttr(sds_id, 'valid_range');
valid_range = sd.readAttr(sds_id, valid_range_index);

% Read units.
units_index = sd.findAttr(sds_id, 'units');
units = sd.readAttr(sds_id, units_index);

% Read scale_factor.
scale_index = sd.findAttr(sds_id, 'scale_factor');
scale = sd.readAttr(sds_id, scale_index);
scale = double(scale);

% Read add_offset.
offset_index = sd.findAttr(sds_id, 'add_offset');
offset= sd.readAttr(sds_id, offset_index);
offset = double(offset);

% Terminate access to the dataset.
sd.endAccess(sds_id);

% Close the file.
sd.close(SD_id);


% Replace the fill value with NaN.
data(data==fillvalue) = NaN;

% Replace values outside of the valid_range with NaN.
data(data<valid_range(1)) = NaN;
data(data>valid_range(2)) = NaN;

% Multiply scale and add offset.
data = scale*data + offset;

% Plot data.
pole=[-90 0 0];
latlim=[-90, ceil(max(max(lat)))];
lonlim=[floor(min(min(lon))),ceil(max(max(lon)))];
min_data=floor(min(min(data)));
max_data=ceil(max(max(data)));

f = figure('Name', FILE_NAME, 'visible', 'off');

axesm('MapProjection','stereo', ...
      'MapLatLimit', latlim, 'MapLonLimit', lonlim, ...
      'Origin', pole, 'Frame', 'on', 'Grid', 'on', ...
      'MeridianLabel', 'on', 'ParallelLabel', 'on')
   
surfm(lat, lon, data);
colormap('Jet');
h = colorbar();

% Plot coast lines.
coast = load('coastlines.mat');
plotm(coast.coastlat, coast.coastlon, 'k')

title({FILE_NAME;long_name}, 'Interpreter', 'None', ...
    'FontSize', 10, 'FontWeight', 'bold');

set (get(h, 'title'), 'string', units, 'Interpreter', 'None', ...
                   'FontSize',10,'FontWeight','bold');
saveas(f,[FILE_NAME '.m.png']);
exit;
